src/client/ws/client.rs

```diff
@@ -257,164 +255,205 @@ impl<Notif> Drop for Subscription<Notif> {
-			Either::Right(Ok(RawClientEvent::Response { request_id, result })) => {
-				log::trace!("[backend] client received response to req={:?}, result={:?}", request_id, result);
-				match ongoing_requests.remove(&request_id) {
-					Some(r) => {
-						if let Err(e) = r.send(result.map_err(Error::Request)) {
-							log::error!("Could not dispatch pending request ID: {:?}, error: {:?}", request_id, e);
+				Some(Ok(jsonrpc::Response::Single(response))) => {
+					if let Some((unsubscribe, params)) = process_response(&mut manager, response, config.subscription_channel_capacity) {
+						sender.start_request(unsubscribe, params).await.unwrap();
+					}
+				}
+				Some(Ok(jsonrpc::Response::Batch(responses))) => {
+					for response in responses {
+						if let Some((unsubscribe, params)) = process_response(&mut manager, response, config.subscription_channel_capacity) {
+							sender.start_request(unsubscribe, params).await.unwrap();
 						}
 					}
-					None => log::error!("No pending response found for request ID {:?}", request_id),
 				}
-			}
-
-			// Received a response from the server that a subscription is registered.
-			Either::Right(Ok(RawClientEvent::SubscriptionResponse { request_id, result })) => {
-				log::trace!("[backend]: client received response to subscription: {:?}", result);
-				let (send_back, unsubscribe) = pending_subscriptions.remove(&request_id).unwrap();
-				if let Err(err) = result {
-					let _ = send_back.send(Err(Error::Request(err)));
-				} else {
-					let (notifs_tx, notifs_rx) = mpsc::channel(config.subscription_channel_capacity);
-
-					// Send receiving end of `subscription channel` to the frontend
-					if send_back.send(Ok((notifs_rx, request_id))).is_ok() {
-						active_subscriptions.insert(request_id, (notifs_tx, unsubscribe));
-					} else {
-						close_subscription(&mut client, request_id, unsubscribe).await;
+				Some(Ok(jsonrpc::Response::Notif(notif))) => {
+					let sub_id = notif.params.subscription;
+					let request_id = match manager.get_request_id_by_subscription_id(&sub_id) {
+						Some(r) => r,
+						None => {
+							log::error!("Subscription ID: {:?} not found", sub_id);
+							continue;
+						}
+					};
+
+					match manager.as_active_subscription_mut(&request_id) {
+						Some(send_back_sink) => {
+							if let Err(e) = send_back_sink.try_send(notif.params.result) {
+								log::error!("Dropping subscription {:?} error: {:?}", sub_id, e);
+								manager.remove_active_subscription(request_id, sub_id).unwrap();
+							}
+						}
+						None => {
+							log::error!("Subscription ID: {:?} not an active subscription", sub_id);
+						},
 					}
 				}
+				Some(Err(e)) => {
+					log::error!("error: {:?} terminating client", e);
+					return;
+				}
+			},
+		}
+	}
+}
+
+/// Process a response for the Server.
+///
+/// Returns Some(method, params) if the client has to send back any response.
+/// None otherwise.
+fn process_response(
```

Comment:
I just realized that this should probably return `Result<Option<callback>, Err>` if the request was not found it need to be propagated and handled to terminate the `background thread` because `Err(e)` from the backend doesn't check request id anymore.

It's one if this function's tasks.